### PR TITLE
Add complex form binding support to RDF

### PR DESCRIPTION
**Pull request by captainsafia**
- Add trimming annotations to shared form binding code
- Add basic support for form-binding for complex types to RDF

```csharp
var app = WebApplication.Create();

app.MapPost("/", ([FromForm] Todo todo) => Results.Ok(todo));

app.Run();

record Todo(int Name, bool IsCompleted);
```

### CODE REVIEW

**Comment by mitchdenny** [comment]
Is RDG support coming later?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -730,8 +737,14 @@ private static Expression CreateArgument(ParameterInfo parameter, RequestDelegat
                 }
                 return BindParameterFromFormCollection(parameter, factoryContext);
             }
-
-            return BindParameterFromFormItem(parameter, formAttribute.Name ?? parameter.Name, factoryContext);
+            var useSimpleBinding = parameter.ParameterType == typeof(string) ||
```
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:740]: Nit: comment on why this is the criteria used for simple vs complex binding?

**Comment by captainsafia** [comment]
> Is RDG support coming later?

Yep. The meta issue for this is available at https://github.com/dotnet/aspnetcore/issues/48759. "Converter source generation" outlines the requirement to codegen implementations of the converters for the serializer that are reflection-free so that we can use them in RDG.

I've done some prototyping in this area but wanted to get this MVP out for RDF to make some movement in the space.

File: src/Components/Endpoints/src/Binding/FormBindingHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.Endpoints.Binding;
+
+internal static class FormBindingHelpers
+{
+    public const string RequiresUnreferencedCodeMessage = "Form binding is not trim compatible.";
```
Comment (adityamandaleeka) [inline src/Components/Endpoints/src/Binding/FormBindingHelpers.cs:8]: Is this more helpful?

```suggestion
    public const string RequiresUnreferencedCodeMessage = "Form binding is not compatible with trimming, as it requires dynamic access to code that is not referenced statically.";
```

**Comment by mitchdenny** [comment]
Gotta start somewhere. I'm a little concerned that we'll have diverging support for these kinds of features in RDF and RDG which is why I picked it up.

File: src/Components/Endpoints/src/Binding/FormBindingHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    public const string RequiresDynamicCodeMessage = "Form binding may require dynamic code gen.";
```
Comment (adityamandaleeka) [inline src/Components/Endpoints/src/Binding/FormBindingHelpers.cs:9]:
```suggestion
    public const string RequiresDynamicCodeMessage = "Form binding may require dynamic code generation.";
```

**Review by adityamandaleeka — APPROVED** [review]
Left some minor nit comments but LGTM.

**Review by MackinnonBuck — APPROVED** [review]
Looks great!

**Pull request merged**